sdk/tables/azure-tables/src/EdmInt64.ts

```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ * A class that allows you to represent a 64-bit signed integer as a string.
```

Comment:
what does this class really do? It seems to just be a box around a string, it doesn't actually appear to help you work with the value as a number?

Reply:
It's possible that we could add some function that would allow you to know if the underlying integer would fit safely into a `number` value (i.e. is less that Number.MAX_SAFE_INTEGER) and a way to get it out as a `number` when that holds?  I suspect that in practice if the number here is larger that it won't be easy to work with in in JavaScript (you'd likely need to construct a `BigInt` from it, assuming that your runtime had support for `BigInt`.

Reply:
@xirzec the main goal of this class is to provide "metadata" about the value. The service would return something like this:

```json
{
   "foo": "255486129307",
   "foo@odata.type": "Edm.Int64",
   "bar": "a4045b38-33f3-4fe8-8661-bd906c76531b",
   "bar@odata.type": "Edm.Guid"
}
```

In order to provide consumers a nicer object we want to deserialize it to 

```typescript
interface MyEntity {
   foo: EdmInt64;
   bar: EdmGuid;
}
```

Similarly when sending an entity to the service, instead of having to build their objects with the extra `propertyName@odata.type` properties, users would be able to express that with these classes.


```typescript

const myEntity = {
   foo: new EdmInt64("255486129307"),
   bar: new EdmGuid("a4045b38-33f3-4fe8-8661-bd906c76531b")
}
```

And the serializer will translate that into the `odata.type` form.


@ellismg we thought about trying not to provide much additional functionality to avoid giving users the impression that this would provide any kind of big number handling. However having a isSafeInteger may be useful, we could add a function that just calls `Number.isSafeInteger(this.value)`, and users can decide what to use if they want to do arithmetic with the value. Sounds good?

Reply:
```typescript
public isSafeInteger() {
  const numericValue = Number(this.value);
  return Number.isSafeInteger(numericValue)
}
```

Reply:
Maybe it's worthwhile?  I could imagine not having this wouldn't be the end of the world.  In the case where `isSafeInteger` here returned `true`, the next thing you'd do would be something like `const num = Number(entity.value)` anyway, and so you could image that maybe it would be reasonable to just force folks to create the number object and then call `isSafeInteger` on it?

Reply:
I guess this just feels very .NET to me - having a "type" for  GUID even though it's really just a string over the wire in the JSON. In general we don't abstract over-the-wire types except for Dates since those are a lot more usable for the programmer than a number or string.

Reply:
To give a background on the reason we are adding these: Here's an [overview of the types that are natively supported by the Tables Service](https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-the-table-service-data-model).  When we submit entities, if we want their data-type in the table to be an `int64` or a `GUID` we need to add additional metadata onto the entity to denote the type of the property.  And when we read existing entities, this information will be returned to us.  In a pure JavaScript world, you're right that this feels a little weird.  The thing we are trying to support is being able to reflect these data types for the entity fields in cases where that's important, which will usually be cross language scenarios (i.e. I want to write an entity from JavaScript and that entity contains a GUID or Int64 and I want to mark it in a way such that when it's consumed by .NET or Java or Python it is converted to the correct native type instead of just being a "string").

IIRC, in Track 1, you would accomplish this via an object that fit a shape like `{ edmType: "Edm.Int64" | "Edm.Guid" | ... , value: string }`.  The goal here was to make things a little nicer, but maybe we've missed the mark?

I do think that we need to have a way to denote you want to use these types which have no analog in the JavaScript type system (which is more or less Int64 and UUID (and you can squint a little and think maybe we'll have to do something special for Int32 as well, at some point)).

Does the overall scenario make sense?

Reply:
The Int64 problem makes sense to me, ideally we support BigInt for this, but having some way to mark a string as "really" being an Int64 is perhaps useful.

I see the problem with Guid, since when consuming it's really just an opaque string, but how would someone uploading data to the system denote it without some nominal type? The strategy we used for Search would say to regex test the string to see if it looked like a Guid and if it did, treat it as such.

The type approach is much less ambiguous and lets the developer control it... but realistically if they generated a Guid in JS they did so as a string (probably with the "uuid" package) and now you're saying they'd have to remember to take their string and convert it by hand over to use your special Edm.Guid type... I'm not sure that they'd discover that very easily.

The other issue is that since this type is local to the package, if we later had a different package (e.g. storage-blob) that wanted to have an Edm.Guid type, it wouldn't be the same and instanceof checks would fail.

Reply:
> The Int64 problem makes sense to me, ideally we support BigInt for this

Quick Question - What's our general position here with respect to how far back of Node we want to support?  `BigInt` support landed in Node 10.4.0, which was an LTS release.  The previous LTS release, Node 8 has been end of life'd, but I know in other packages we still say Node 8 above are supported.

Would we be willing to use `BigInt` here somehow?  CanIUse shows reasonable browser support (though for Safari it looks like it's only present in technical previews).

Reply:
My feeling is that wrapper types for guids and bigints are a bridge too far, but I can't say I follow the cross-language round-tripping constraint.

We support Node 8 and myriad browsers that don't support bigints. Unfortunately we can't depend on them. Even more unfortunately, they are highly resistant to polyfilling.

Reply:
@xirzec @bterlson @ellismg , if we decide we don't want to use EdmGuid, we'd have 3 options when serializing right?

(1) Just treat Guid as strings, don't do anything special
(2) Regex test all the strings and add `propertyName@odata.type: 'Edm.Guid'` when the test is true
(3) Users explicitly set `propertyName@odata.type: "Edm.Guid"`

**(1)** is the simpler approach since neither we or the users need to do anything. With this route we'd ignore Tables Edm.Guid 
 also the problem I see with this approach is that we may find conflicts when interacting with other languages. For example, if we insert an entity 

```typescript
const entity = {id: "07b5cbfa-ed0c-4a5e-a372-534476aa4a05", name: "TestEntity"}

/* Serialized as:
*
*  {
*    "id": "07b5cbfa-ed0c-4a5e-a372-534476aa4a05", 
*    "id@odata.type": "Edm.String", 
*    "name": "TestEntity",
*    "name@odata.type": "Edm.String", 
* }
*/
```

And C# tries to consume it, it would be expecting `"id@odata.type": "Edm.Guid"`

**(2)** With this approach we'll try to be smart and assume that if something looks like a Guid it is a Guid, I think this may be fine most of the times, this would be a bit inconvenient for scenarios where you actually want a string but your string happens to have the shape of a Guid as it would end up stored as a GUID.


**(3)** Gives complete control to the user, we'll just handle stuff we can safely infer its type. The other types are responsibility of the user. This approach is explicit enough to guarantee the right types when needed, the downside is that the user experience is not as smooth.

```typescript

// One way to represent non-native types, using an object when you need to send "special types" such as Guid and Int64
const myEntity = {
   id: { value: "07b5cbfa-ed0c-4a5e-a372-534476aa4a05", type: "Edm.Guid"},
   name: "Test Entity"
}

// Another option closer to the RestApi
const myOtherEntity = {
   id: "07b5cbfa-ed0c-4a5e-a372-534476aa4a05",
   "id@odata.type": "Edm.Guid",
   name: "Test Entity"
}
```


The reason we thought about the wrapper classes is that they seemed to be a middle ground between being able to explicitly represent types and user experience.


What do you guys like better? Also I may be missing other options so please feel free to suggest.

I like the wrappers since we can also add validations, my second choice would be **(3)**.

By the way, I'm really happy to be having this conversation, thanks guys for bringing this up :smile: